### ISSUE TITLE
Complete implementation of huabao protocol (JT808 document attached)

### ISSUE THREAD

**Issue by K1DV5** [traccar/traccar#4871]
I'm using the huabao protocol and I need to get fuel consumption reports. I checked if it is implemented but many things are not implemented for this protocol, including fuel consumption. I checked the linked documents in https://www.traccar.org/protocols/ and those documents are not fully JT808 compliant. I have the full document [here](https://drive.google.com/file/d/1UE3jKSM-hTrcS88DEHAJIukxcJgqYBQF/view?usp=sharing). But I have little to no java knowledge. Can you please implement the full location information report protocol, including the fuel level part?

**Comment by K1DV5** [comment]
I understand. And I apologize for not checking thoroughly, it seems that fuel level and consumption is implemented. But there are still features not implemented, especially many commands. Will leave it here as you said.

**Comment by K1DV5** [comment]
Hi Anton, I want to slowly implement this (and learn java in the process.) I want to start with canbus data uploading (0x0705) and telephone handling (0x8400 and 0x8401). Do you have any suggestion on where I should start?

**Comment by tananaev** [comment]
Check the decoder class. That's where all the relevant code is.

**Comment by K1DV5** [comment]
Thank you @tananaev. I'm looking at the class and one question came up. The can bus data comes on its own, not attached to the location report. It is a report on its own, with its own interval. So how can I handle that? Should I make it an event? Or something else?

**Comment by K1DV5** [comment]
Thank you, will continue working on #4904.

**Comment by K1DV5** [comment]
Hi Anton, I implemented 0x0705 (can bus data decoding), wrote a test which passed and all but when I tried it with a real device, I get this log entry:

```
2022-07-25 14:26:04  WARN: Failed to store position - ERROR: value too long for type character varying(4000) - PSQLException (... < QueryBuilder:472 < DatabaseStorage:89 < DefaultDataHandler:45 < ...)
```
What I did was decode the data into a list, which can contain several items (frames), convert it into a json string and store it as an attribute using `Position.set()`. I think the problem is that the attributes column in the database has a type `character varying(4000)` and the data is more than that. How can I tackle this?

**Comment by K1DV5** [comment]
@tananaev can I just change the column type to `character varying` (without a limit)? That would change it for all protocols but I don't think it will break anything.

**Comment by tananaev** [comment]
You can change it locally, but we want to have some limit in the official version.

**Comment by K1DV5** [comment]
Ok can we make it bigger? Like 100k?

**Comment by tananaev** [comment]
No because it doesn't work on all databases. We already selected the highest limit.

**Comment by K1DV5** [comment]
Ok so I'll save it across multiple positions as a workaround, although I fear it'll perform a little worse and bloat the database.

Another question, is it ok if lists and maps are accepted as position attributes? Because although I can just convert it to a json string (which is what I do now), I think it'll be redundant as it will later be saved as json among the other attributes. I'm asking to add overloads for `Position.set` that'll accept lists and maps.

**Comment by tananaev** [comment]
Can you please provide more details on what you're trying to save exactly. It seems weird that 4000 characters is not enough.

**Comment by K1DV5** [comment]
Right, the device sends [CAN bus data]( https://www.csselectronics.com/pages/can-bus-simple-intro-tutorial), good info about the vehicle, in a format most of which is a sequence of 4 byte `can id` and 8 byte `can data`.
```
58f64544 576d358c44365497
58d24344 56654b6557733475
...
```
The first 3 bits of `can id` is info about channel, originality and frame type. The rest 29 bits are the frame ID. Then the data follows. The device sends several frames at once and that's where the problem happened. I was trying to store the data as a list of maps.

**Comment by tananaev** [comment]
Yeah, you should probably split frames.

**Comment by nilei** [comment]
@tananaev and @K1DV5

Is this issue fixed or any related PR merged? I can see a trend in mature devices using the JT808 protocol. 

-- Cheers, Nils

**Comment by tananaev** [comment]
What trend?

**Comment by nilei** [comment]
> What trend?

Hi Anton,

Great to finally connect with you directly! 🙌

**Context:** When I mention JT808, I'm really referring to the entire class of Chinese manufacturer protocols—and that's where the market currently lives. A quick reality check: post an RFQ for 4G GPS trackers on Alibaba, and you'll get ~10 responses in your inbox. All of them support JT808. Not an edge case—it's the industry standard.

**Why this matters now:** We're at the inflection point from reactive trackers to predictive IoT devices (AI-powered forecasting instead of just location pings). Chinese manufacturers are driving this shift aggressively—and they're standardizing on JT808. Excluding these devices means missing the largest innovation pool in the industry.

**What's in it for Traccar:**
✅ Even broader hardware coverage without fragmented protocol implementations
✅ Future-proofing through support of AI-capable devices
✅ Active community support — I'm happy to provide test devices and testers

**Here's a concise market trend summary:** [Google AI Mode Summary](https://share.google/aimode/hlIplo000ie1wDdxx)

What's your take—does it make sense to prototype this together?

— Cheers, Nils

**Comment by tananaev** [comment]
What is not working for you?

**Comment by nilei** [comment]
Hi @tananaev ,

I may have been unclear—my question is straightforward: **Is JT808 currently supported, or is support planned?**

I don't have a technical issue to debug; I simply need to know whether I can commit to JT808 hardware for an upcoming deployment, or if I should pursue alternatives.

A **gentle observation**: Your brief one-liners can sometimes come across as **abrupt**—especially to community members genuinely motivated to contribute. I'm certain that's not your intent, but it can dampen the collaborative exchange that moves Traccar forward. A bit more context in responses would go a long way here.

I genuinely value your time and the work you put into the project. Looking forward to your thoughts.

— Cheers, Nils

**Comment by tananaev** [comment]
It has been supported for a very very long time.

**Comment by nilei** [comment]
> It has been supported for a very very long time.

Hi Anton,

thanks for clarifying—great to hear JT808 is already supported!

That said, I checked https://www.traccar.org/protocols/ before reaching out, and JT808 isn't listed there. That's likely why I (and presumably others) assumed it wasn't implemented yet.

Quick suggestion: Adding it to the protocols page would prevent future confusion and save you from repetitive questions like mine.

Appreciate the quick response!

— Cheers, Nils

**Comment by tananaev** [comment]
It's called Huabao.

**Comment by nilei** [comment]
> It's called Huabao.

Hi Anton,

ah, got it—**Huabao** it is. Thanks for pointing that out.

**Small suggestion**: Since most manufacturers and documentation refer to it as JT808 (and many users search for that term), adding "JT808" as an alias or note on the protocols page could help people like me find it faster. Would save you from these repetitive back-and-forths.

Appreciate the patience!

— Cheers, Nils

**Comment by tananaev** [comment]
It's already added.

**Comment by nilei** [comment]
> It's already added.

Hi Anton,

strange—I just checked https://www.traccar.org/protocols/ again and searched for "JT808", but it does not show up in the list or search results.

Could be a caching issue, or perhaps I'm looking at the wrong section?

— Cheers, Nils

**Comment by tananaev** [comment]
It's here:

https://www.traccar.org/devices/

**Comment by nilei** [comment]
Hi Anton,

got it—thanks for the pointer. I can see it's listed as JT808 (port 5015) on the devices page.

**Just a heads-up**: since JT808 is the actual protocol name (Huabao being one implementation/manufacturer), and that's what most Chinese manufacturers and users search for, adding "JT808" as a cross-reference or alias on the protocols/devices page could prevent future confusion. I searched for "JT808" initially and came up empty, which led to this thread.

Not a big deal—just trying to save you from similar questions down the line.

— Cheers, Nils

**Comment by tananaev** [comment]
It's listed as the device name because there's no plan to rename the protocol.

**Comment by nilei** [comment]
Hi Anton,

fair enough—thanks for clarifying the naming approach. Good to know JT808/Huabao is supported and here to stay.

Appreciate your patience with my questions. Closing this now.

— Cheers, Nils

**Comment by tananaev** [comment]
Closing as no pending actions here. If something is still missing, please open a new ticket.
